From f8e0c6837511fa9421526f9c59d02ab17d612fc4 Mon Sep 17 00:00:00 2001
From: Raj Kamal <rkamal@codeaurora.org>
Date: Mon, 3 Mar 2014 23:55:33 +0000
Subject: [PATCH 6/6] gralloc: Perform func to return stride and height

Backport from CM12 commit 11349e1dd46e6fee2a6f82667d15d811d8a3f378.
Adapt msm8084 paths to CM11 display-caf and reject null output pointers.

Change-Id: I44c814b2bae1f61d3ec5c36a22d1c6119542b10a
---
 libgralloc/gralloc_priv.h |  1 +
 libgralloc/mapper.cpp     | 18 ++++++++++++++++++
 2 files changed, 19 insertions(+)

diff --git a/libgralloc/gralloc_priv.h b/libgralloc/gralloc_priv.h
index 58fc898..ab686e5 100644
--- a/libgralloc/gralloc_priv.h
+++ b/libgralloc/gralloc_priv.h
@@ -81,6 +81,7 @@ enum {
     GRALLOC_MODULE_PERFORM_CREATE_HANDLE_FROM_BUFFER = 1,
     GRALLOC_MODULE_PERFORM_GET_STRIDE,
     GRALLOC_MODULE_PERFORM_GET_CUSTOM_STRIDE_FROM_HANDLE,
+    GRALLOC_MODULE_PERFORM_GET_CUSTOM_STRIDE_AND_HEIGHT_FROM_HANDLE,
 };
 
 #define GRALLOC_HEAP_MASK   (GRALLOC_USAGE_PRIVATE_UI_CONTIG_HEAP |\
diff --git a/libgralloc/mapper.cpp b/libgralloc/mapper.cpp
index cfe6dbf..1f97524 100644
--- a/libgralloc/mapper.cpp
+++ b/libgralloc/mapper.cpp
@@ -383,6 +383,24 @@ int gralloc_perform(struct gralloc_module_t const* module,
                 }
                 res = 0;
             } break;
+        case GRALLOC_MODULE_PERFORM_GET_CUSTOM_STRIDE_AND_HEIGHT_FROM_HANDLE:
+            {
+                private_handle_t* hnd = va_arg(args, private_handle_t*);
+                int *stride = va_arg(args, int *);
+                int *height = va_arg(args, int *);
+                if (private_handle_t::validate(hnd) || !stride || !height) {
+                    break;
+                }
+                MetaData_t *metadata = (MetaData_t *)hnd->base_metadata;
+                if (metadata && metadata->operation & UPDATE_BUFFER_GEOMETRY) {
+                    *stride = metadata->bufferDim.sliceWidth;
+                    *height = metadata->bufferDim.sliceHeight;
+                } else {
+                    *stride = hnd->width;
+                    *height = hnd->height;
+                }
+                res = 0;
+            } break;
         default:
             break;
     }
-- 
2.53.0

